### PR TITLE
Introduce checks to validate public API tests

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131390

This PR introduces a new sanity check for the public API tests in `.ci/pytorch/test.sh`.
* Validates two public API tests:
    1. Ensures `test_correct_module_names` fails when a new file OR an existing file adds an invalid public API function (e.g. one whose `__module__` is unset).
    2. Ensures `test_modules_can_be_imported` fails when a module underneath `torch/` cannot be imported.
* Runs this in CI as part just before the pre-existing FC / BC checks.

I've verified that re-introducing the bug that #131386 fixed causes the new check to fail:
public_api_failure

### CODE REVIEW

File: test/test_public_bindings.py [hunk 1]
```diff
@@ -594,7 +594,8 @@ def check_one_element(elem, modname, mod, *, is_public, is_all):
                         )
 
         for mod in pkgutil.walk_packages(torch.__path__, "torch."):
-            modname = mod.name
+            # TODO: Change this back after verifying the check catched the bug in CI
```
Comment (jbschlosser) [inline test/test_public_bindings.py:597]: catched -> caught lmao

File: test/forward_backward_compatibility/check_public_API_tests.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            "test/test_public_bindings.py for any changes that may have broken the test."
+        )
+        sys.exit(1)
+
+    print(
+        "Success! 'test_correct_module_names' identified a new, invalid public API function."
+    )
+
+# Step 2. Make sure that the public API test "test_modules_can_be_imported" fails when an existing
+# file is modified to introduce an invalid public API function.
```
Comment (albanD) [inline test/forward_backward_compatibility/check_public_API_tests.py:61]: Huh, this test checks that "import your.module" doesn't crash :p 
What you describe here would be running test_correct_module_names again but after modifying an existing file right?

Comment (jbschlosser) [reply test/forward_backward_compatibility/check_public_API_tests.py:61]: Hm I'm trying to follow the description for this task:
```
In particular we want to make sure that both these tests run and are not broken (this happened before during some CI infra update). These tests can be (and have been) silently broken/disabled before and we have no signal about it! A way to do this is to have the lint shard run the following: 1. create a bad file 2. make sure test_public_bindings.py -k test_correct_module_names fails 3. append bad API to existing file 4. make sure test_modules_can_be_imported (same file) fails 5. undo the changes.
```

In practice, I've seen that modifying `torch/nn/functional.py` with an invalid public function does indeed fail `test_modules_can_be_imported`, but do we care about testing this test and, if so, if this the right way to do it?

Comment (jbschlosser) [reply test/forward_backward_compatibility/check_public_API_tests.py]: Updated to include the following test cases:
1. test_correct_module_names() fails for an invalid public API function in a new file
2. test_correct_module_names() fails for an invalid public API function in an existing file
3. test_modules_can_be_imported() fails when some module can't be imported

File: test/forward_backward_compatibility/check_public_API_tests.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        f.seek(pos, os.SEEK_SET)
+        f.truncate(pos)
+
+# Step 3. Make sure that the public API test "test_modules_can_be_imported" fails when a module
+# cannot be imported.
+with tempfile.TemporaryDirectory(dir="torch") as d:
+    print("Creating a non-importable module...")
+    with open(os.path.join(d, "__init__.py"), "wb") as f:
+        f.write(b"invalid syntax garbage")
+        f.seek(0)
```
Comment (albanD) [inline test/forward_backward_compatibility/check_public_API_tests.py:103]: nit: not needed

**Review by albanD — APPROVED** [review]
SGTM!

**Review by albanD — APPROVED** [review]
SGTM !

**Pull request closed without merging**